src/Xmobar/Config/Types.hs

```
@@ -91,3 +95,12 @@ data Border = NoBorder
             | BottomBM Int
             | FullBM Int
               deriving ( Read, Show, Eq )
+
+newtype SignalChan = SignalChan { unSignalChan :: Maybe (STM.TMVar SignalType) }
+
+instance Read SignalChan where
+  readsPrec _ s = [ (SignalChan Nothing, s) ]
```

Why should there be a real read instance?  Maybe I misunderstand, but it seems like using the channel require access to the `TMVar` which should come from the client?  That is, am I correct in assuming that an empty signal channel read from a config is unusable for calling code?

---

I left the empty Read instance so that users who compile xmobar and spawn it as a separate os process or have a 'standard' configuration file don't end up making a TMVar that they never use.  Really this is for people using xmonad or some other haskell wm who want to use xmobar more like a library.

---

An alternative I played with was to accept `Maybe TMVar` in `xmobar` as a second parameter and that removes the need to have it in the config but it doesn't feel quite as nice for the calling code

---

in other words, if you are using xmobar from an xmonad.hs then you can use the signal channel, otherwise, xmobar should create one and do the normal execution with os signal handling setup and such.

---

The convention that I followed when I made the QueueReader plugin was a failing Read instance which I copied from `HandleReader`:

``` haskell
-- | WARNING: This read instance will throw an exception if used! It is
-- only implemented, because it is required by 'Xmobar.Run` in 'Xmobar.commands'.
instance Read (QueueReader a) where
  -- | Throws an 'error'!
  readsPrec = error "QueueReader: instance is a stub"
```

While this still feels hacky, I think it is better to fail loud and hard if someone tries to add these unreadable plugins to a text config.

---

Using the failing `Read` also makes it clear that the `STM` object doesn't belong in the config datatype.

---

Hmm, okay it does still seem wrong to me to have read instance for the signal channel do anything but fail, since it should never be used as such.